src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+                _runLoopStartedEvent.Set();
+
+                // Start the OS X RunLoop (a blocking call) that will pump file system changes into the callback function
+                Interop.CoreFoundation.CFRunLoopRun();
+
+                // When we get here, we've requested to stop so cleanup the EventStream and unschedule from the RunLoop
+                Interop.CoreServices.FSEventStreamStop(_eventStream);
+                Interop.CoreServices.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.CoreFoundation.kCFRunLoopDefaultMode);
+            }
+            else
+            {
+                // An error occurred while trying to start the run loop so fail out
+                Interop.CoreServices.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.CoreFoundation.kCFRunLoopDefaultMode);
+                OnError(new ErrorEventArgs(new IOException("", Marshal.GetLastWin32Error())));
+            }
+        }
+
+        private void FileSystemEventCallback( 
+            FSEventStreamRef streamRef, 
+            IntPtr clientCallBackInfo, 
+            size_t numEvents, 
+            String[] eventPaths, 
+            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
+            Interop.CoreServices.FSEventStreamEventFlags[] eventFlags,
+            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
+            FSEventStreamEventId[] eventIds)
+        {
+            Debug.Assert((numEvents.ToInt32() == eventPaths.Length) && (numEvents.ToInt32() == eventFlags.Length) && (numEvents.ToInt32() == eventIds.Length));
+
+            // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll add it to this
+            // list so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
+            System.Collections.Generic.List<FSEventStreamEventId> handledRenameEvents = new Collections.Generic.List<FSEventStreamEventId>();
+
+            for (long i = 0; i < numEvents.ToInt32(); i++)
+            {
+                // First, we should check if this event should kick off a re-scan since we can't really rely on anything after this point if that is true
+                if (ShouldRescanOccur(eventFlags[i]))
+                {
+                    OnError(new ErrorEventArgs(new IOException("", (int)eventFlags[i])));
+                    break;
+                }
+                else if (handledRenameEvents.Contains(eventIds[i]))
+                {
+                    // If this event is the second in a rename pair then skip it
+                    continue;
+                }
+                else
+                {
+                    // Check if the path passes the name filter and if the event passes the event filter.
+                    if ((DoesPathPassNameFilter(eventPaths[i])) && (_filterFlags.HasFlag(eventFlags[i])))
```

Comment:
Nit: does this need to be a nested if as it is, or could you combine it with the previous else to be an else if?